src/Control/Effect/Interpret.hs

```diff
@@ -23,27 +77,27 @@ import Control.Monad.Trans.Class
 --   At time of writing, a simple passthrough use of 'runInterpret' to handle a 'State' effect is about five times slower than using 'StateC' directly.
 --
 --   prop> run (runInterpret (\ op -> case op of { Get k -> k a ; Put _ k -> k }) get) === a
-runInterpret :: (forall x . eff m x -> m x) -> InterpretC eff m a -> m a
-runInterpret handler = runReader (Handler handler) . runInterpretC
-
-newtype InterpretC eff m a = InterpretC { runInterpretC :: ReaderC (Handler eff m) m a }
-  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadFix, MonadIO, MonadPlus)
+runInterpret
+  :: forall eff m a.
+     ( HFunctor eff, Monad m )
+  => ( forall x. eff m x -> m x )
+  -> ( forall s. Reifies s ( Handler eff m ) => InterpretC s eff m a )
+  -> m a
+runInterpret f m =
+  reify ( Handler handler ) ( go m )
 
-instance MonadTrans (InterpretC eff) where
-  lift = InterpretC . lift
-  {-# INLINE lift #-}
+  where
 
-newtype Handler eff m = Handler (forall x . eff m x -> m x)
+    handler :: forall s x. eff ( InterpretC s eff m ) x -> InterpretC s eff m x
+    handler e =
+      InterpretC ( f ( handleCoercible e ) )
 
-runHandler :: (HFunctor eff, Functor m) => Handler eff m -> eff (InterpretC eff m) a -> m a
-runHandler h@(Handler handler) = handler . hmap (runReader h . runInterpretC)
-
-instance (HFunctor eff, Carrier sig m) => Carrier (eff :+: sig) (InterpretC eff m) where
-  eff (L op) = do
-    handler <- InterpretC ask
-    lift (runHandler handler op)
-  eff (R other) = InterpretC (eff (R (handleCoercible other)))
-  {-# INLINE eff #-}
+    go
+      :: forall x s.
+         InterpretC s eff m x
+      -> Tagged s ( m x )
```

Comment:
If we don’t quantify the variables in `runInterpret`’s signature, then we don’t need to quantify the variables here either.

Reply:
I am not quantifying everything, only `x` and `s`. I need to share `m`. This can't be written without scoped type variables/inlining definitions.